### ISSUE TITLE
F31: openQA live image build test failing when update set includes a kernel

### ISSUE THREAD

**Issue by AdamWill** [weldr/lorax#987]
Filing this to track a pretty mysterious bug that's been causing me grief with the openQA update tests over the last few weeks.

We have this one openQA test that [builds a live image](https://pagure.io/fedora-qa/os-autoinst-distri-fedora/blob/master/f/tests/_live_build.pm). It is run as part of update testing (where we test updates from updates-testing). The whole flow of the test goes like this:

1. Boot from a pre-rolled disk image that's basically a minimal install of the release
2. Set up a side repository containing packages from the update under test (this involves installing the koji and bodhi CLI clients, git, and createrepo)
3. Run `dnf -y update` to update the system with both the latest packages from stable updates, and the packages from the update under test
4. Reboot the system
5. Create a live image by running all the commands you can see in the test code linked above; the image is built in mock, the process is intended to follow how lives are built in Koji as closely as possible without actually setting up a Koji instance. The side repo containing the update packages **is** enabled in both the mock configuration and the kickstart used for the live image build process

it seems that, **when the update set in steps 2/3 includes a kernel** - i.e. either there's a kernel package in the update under test, or there's a newer kernel in stable updates than in the base disk image - the live image build often or always fails due to the `install.img` file somehow disappearing once the process gets to the stage of running templates, after the anaconda run has completed. The error looks like this:

    2020-04-08 19:36:21,345 INFO pylorax: Building boot.iso
    2020-04-08 19:36:21,353 DEBUG pylorax.treebuilder: kernels=[{'path': 'boot/vmlinuz-5.5.16-200.fc31.x86_64', 'version': '5.5.16-200.fc31.x86_64', 'arch': 'x86_64', 'flavor': None, 'initrd': {'path': 'boot/initramfs-5.5.16-200.fc31.x86_64.img'}}]
    2020-04-08 19:36:21,354 DEBUG pylorax.ltmpl: executing x86.tmpl with variables={'kernels': [{'path': 'boot/vmlinuz-5.5.16-200.fc31.x86_64', 'version': '5.5.16-200.fc31.x86_64', 'arch': 'x86_64', 'flavor': None, 'initrd': {'path': 'boot/initramfs-5.5.16-200.fc31.x86_64.img'}}], 'arch': {'buildarch': 'x86_64', 'basearch': 'x86_64', 'libdir': 'lib64', 'bcj': 'x86'}, 'product': {'name': 'Fedora-Workstation-Live', 'version': '31', 'release': '', 'variant': '', 'bugurl': '', 'isfinal': False}, 'runtime_img': 'images/install.img', 'runtime_base': 'install.img', 'inroot': '/var/tmp/lorax.imgutils.mqet68ks', 'outroot': '/var/tmp/lmc-work-o1ged3gx', 'basearch': 'x86_64', 'libdir': 'lib64', 'isolabel': 'FWL-FEDORA-2020-a188b75f17', 'udev': <function udev_escape at 0x7fecc00a6680>, 'domacboot': True, 'doupgrade': True, 'workdir': None, 'lower': <function string_lower at 0x7fecc00a6710>, 'extra_boot_args': '', 'exists': <function LoraxTemplateRunner.__init__.<locals>.<lambda> at 0x7fecbf993170>, 'glob': <function LoraxTemplateRunner.__init__.<locals>.<lambda> at 0x7fecbf8d0f80>}
    2020-04-08 19:36:21,394 ERROR pylorax.ltmpl: Problem rendering x86.tmpl ({'kernels': [{'path': 'boot/vmlinuz-5.5.16-200.fc31.x86_64', 'version': '5.5.16-200.fc31.x86_64', 'arch': 'x86_64', 'flavor': None, 'initrd': {'path': 'boot/initramfs-5.5.16-200.fc31.x86_64.img'}}], 'arch': {'buildarch': 'x86_64', 'basearch': 'x86_64', 'libdir': 'lib64', 'bcj': 'x86'}, 'product': {'name': 'Fedora-Workstation-Live', 'version': '31', 'release': '', 'variant': '', 'bugurl': '', 'isfinal': False}, 'runtime_img': 'images/install.img', 'runtime_base': 'install.img', 'inroot': '/var/tmp/lorax.imgutils.mqet68ks', 'outroot': '/var/tmp/lmc-work-o1ged3gx', 'basearch': 'x86_64', 'libdir': 'lib64', 'isolabel': 'FWL-FEDORA-2020-a188b75f17', 'udev': <function udev_escape at 0x7fecc00a6680>, 'domacboot': True, 'doupgrade': True, 'workdir': None, 'lower': <function string_lower at 0x7fecc00a6710>, 'extra_boot_args': '', 'exists': <function LoraxTemplateRunner.__init__.<locals>.<lambda> at 0x7fecbf993170>, 'glob': <function LoraxTemplateRunner.__init__.<locals>.<lambda> at 0x7fecbf8d0f80>}):
    2020-04-08 19:36:21,409 ERROR pylorax.ltmpl: 

    Traceback (most recent call last):
      File "/usr/lib/python3.7/site-packages/pylorax/ltmpl.py", line 56, in parse
        textbuf = template.render(**variables)
      File "/usr/lib/python3.7/site-packages/mako/template.py", line 476, in render
        return runtime._render(self, self.callable_, args, data)
      File "/usr/lib/python3.7/site-packages/mako/runtime.py", line 883, in _render
        **_kwargs_for_callable(callable_, data)
      File "/usr/lib/python3.7/site-packages/mako/runtime.py", line 920, in _render_context
        _exec_template(inherit, lclcontext, args=args, kwargs=kwargs)
      File "/usr/lib/python3.7/site-packages/mako/runtime.py", line 947, in _exec_template
        callable_(context, *args, **kwargs)
      File "/usr/share/lorax/templates.d/99-generic/live/x86.tmpl", line 22, in render_body
        if os.stat(joinpaths(inroot, runtime_img)).st_size >= 4*1024**3:
    FileNotFoundError: [Errno 2] No such file or directory: '/var/tmp/lorax.imgutils.mqet68ks/images/install.img'

if the update set does **not** include a kernel, this doesn't happen, the live build works fine. So the cycle I'm stuck in is that every few days a kernel update is submitted for F31 (like [this one](https://bodhi.fedoraproject.org/updates/FEDORA-2020-a188b75f17)), the test will [fail](https://openqa.fedoraproject.org/tests/571081) for that update and I can't work around that, then when that update goes stable, the test will start failing for *all* F31 updates until the base disk image is rebuilt (so there is no longer a kernel update in the set of updates from the disk image). So I'm rebuilding the base disk image for F31 manually every few days at this point.

The bug is *not* happening on F30 or F32. I've no idea why not.

I did do a bit of debugging and got as far as figuring out that the problem is *not* that the entire filesystem with the `install.img` on it gets prematurely unmounted, as we kinda suspected. I stuffed some debugging into lorax and it reported that the filesystem was still there, it was really just the *file* that had disappeared. I think the `images/` directory was still present too, but I'm not 100% sure of that any more, I'll have to re-do the debugging if I can find time.

**Comment by AdamWill** [comment]
So I think I figured this out! I don't know why I didn't think of it before, but: I think what's happening is it's running out of disk space in the mock chroot. I hacked up a script to run `df /var/lib/mock` and check for the presence of the image file every few seconds while the build was running and that showed it hitting 100% space used during the image build process and the image file disappearing a bit after that.

So I tried re-running the test with the disk image size bumped 2GB, and it passes :roll_eyes: . Still, it's not at all obvious from the result you get that the actual problem was 'hey you ran out of disk space'.

I am tweaking the test now to actually use a separate 25GB scratch disk image for `/var/lib/mock`, which should hopefully mean we don't run into this problem again for a while. :crossed_fingers: 

Don't know if we just want to close this, or consider whether we can present a more obvious Failure Experience (tm) when disk space is exhausted...
